cpp/src/rolling/grouped_rolling.cu

```diff
@@ -357,15 +357,28 @@ template <typename T>
 struct device_value_accessor {
   column_device_view const col;  ///< column view of column in device
 
+  /// Checks that the type used to access device values matches the rep-type
+  /// of the order-by column.
+  struct is_correct_range_rep {
+    template <typename U>  /// Order-by type.
+    constexpr bool operator()() const
+    {
+      return std::is_same_v<T, cudf::detail::range_rep_type<U>>;
+    }
+  };
+
   /**
    * @brief constructor
    *
    * @param[in] col_ column device view of cudf column
    */
   explicit __device__ device_value_accessor(column_device_view const& col_) : col{col_}
   {
-    cudf_assert(type_id_matches_device_storage_type<T>(col.type().id()) &&
-                "the data type mismatch");
+    // For non-timestamp types, T must match the order-by column's type.
+    // For timestamp types, T must match the range rep type for the order-by column.
+    cudf_assert((type_id_matches_device_storage_type<T>(col.type().id()) or
+                 cudf::type_dispatcher(col.type(), is_correct_range_rep{})) &&
+                "data type mismatch when accessing the order-by column");
```

Comment:
I'm not sure I agree with retrieving the data (via `element()`) using the rep-type and doing operations on said rep-type. I would rather the actual type be read (via `element()`) and used directly.
I feel the original error is correct and this is a flaw in the design of this code perhaps.

Reply:
The reason we're comparing the rep-type values here is that the order-by columns are timestamps, and the preceding/following values are durations (e.g. 2 days ago). 

For integral order-by columns, preceding/following are of the same types. But not for timestamps.

Reply:
Hmm. That said, I could stand to tighten this up a little more... I'll see about a stricter check for timestamps.

Reply:
Ah, I just realized that one unintended consequence of doing order-by comparisons without switching to the rep-types is that all those templates will be instantiated for all timestamp types, duration types, fixed-point, etc. :/